File: server/src/graphs/customer-graph.js
/**
 *
 * 图结构：
 *   START → guardrail → (安全: intentRouter) → (单意图: orderAgent|ragNode|generalChat) → (多意图: Send fan-out) → 所有分支 → answerSynthesizer → END → (拦截: END，跳过后续所有节点)
 */
import { StateGraph, START, END } from "@langchain/langgraph";
import { PostgresSaver } from "@langchain/langgraph-checkpoint-postgres";
import { GraphState } from "./state.js";
import { guardrailNode, routeAfterGuardrail } from "./nodes/guardrail-node.js";
import { intentRouterNode, routeByIntent } from "./nodes/intent-router.js";
import { orderAgentNode } from "./nodes/order-agent.js";
import { ragNode } from "./nodes/rag-node.js";
import { generalChatNode } from "./nodes/general-chat.js";
import { answerSynthesizerNode } from "./nodes/answer-synthesizer.js";
import "dotenv/config";

export const buildCustomerGraph = async () => {
  const connString = `postgresql://${process.env.PG_USER}:${process.env.PG_PASSWORD}@${process.env.PG_HOST}:${process.env.PG_PORT}/${process.env.PG_DATABASE}`;
  const checkpointer = await PostgresSaver.fromConnString(connString);
  await checkpointer.setup();

  const graph = new StateGraph(GraphState)
    .addNode("guardrail", guardrailNode)
    .addNode("intentRouter", intentRouterNode)
    .addNode("orderAgent", orderAgentNode)
    .addNode("ragNode", ragNode)
    .addNode("generalChat", generalChatNode)
    .addNode("answerSynthesizer", answerSynthesizerNode)

    // START → guardrail（安全检查是第一道关卡）
    .addEdge(START, "guardrail")

    // guardrail → (拦截: END) / (放行: intentRouter)
    .addConditionalEdges("guardrail", routeAfterGuardrail, {
      intentRouter: "intentRouter",
      __end__: END,
    })

    // routeByIntent 返回字符串时使用 path map，
    // 返回 Send 数组时绕过 path map 直接 fan-out
    .addConditionalEdges("intentRouter", routeByIntent, {
      orderAgent: "orderAgent",
      ragNode: "ragNode",
      generalChat: "generalChat",
    })

    // 所有分支汇聚到 answerSynthesizer（fan-in）
    .addEdge("orderAgent", "answerSynthesizer")
    .addEdge("ragNode", "answerSynthesizer")
    .addEdge("generalChat", "answerSynthesizer")
    .addEdge("answerSynthesizer", END);

  return graph.compile({ checkpointer });
};

File: backend_py/graphs/customer_graph.py
"""
LangGraph 工作流定义
对照 JS: buildCustomerGraph()
"""
from langgraph.graph import StateGraph, START, END
from graphs.state import GraphState
from graphs.nodes.intent_router import intent_router_node, route_by_intent
from graphs.nodes.order_agent import order_agent_node
from graphs.nodes.rag_node import rag_node
from graphs.nodes.general_chat import general_chat_node
from graphs.nodes.answer_synthesizer import answer_synthesizer_node


def build_customer_graph():
    """
    对照 JS:
      const graph = new StateGraph(GraphState)
        .addNode("intentRouter", intentRouterNode)
        .addNode("orderAgent", orderAgentNode)
        ...
        .addEdge(START, "intentRouter")
        .addConditionalEdges("intentRouter", routeByIntent, {...})
        .addEdge("orderAgent", "answerSynthesizer")
        ...
        .compile()

    Python 差异：方法名从驼峰改为下划线
    """
    graph_builder = StateGraph(GraphState)

    # 注册节点
    graph_builder.add_node("intentRouter", intent_router_node)
    graph_builder.add_node("orderAgent", order_agent_node)
    graph_builder.add_node("ragNode", rag_node)
    graph_builder.add_node("generalChat", general_chat_node)
    graph_builder.add_node("answerSynthesizer", answer_synthesizer_node)

    # 入口：START → 意图识别
    graph_builder.add_edge(START, "intentRouter")

    # 条件路由：意图识别完成后，根据 intent 分流
    graph_builder.add_conditional_edges(
        "intentRouter",
        route_by_intent,
        {
            "orderAgent": "orderAgent",
            "ragNode": "ragNode",
            "generalChat": "generalChat",
        },
    )

    # 三条路径都汇入答案综合节点
    graph_builder.add_edge("orderAgent", "answerSynthesizer")
    graph_builder.add_edge("ragNode", "answerSynthesizer")
    graph_builder.add_edge("generalChat", "answerSynthesizer")

    # 出口
    graph_builder.add_edge("answerSynthesizer", END)

    return graph_builder.compile()